diff --git a/build.gradle b/build.gradle
index 2d948dc..818c335 100644
--- a/build.gradle
+++ b/build.gradle
@@ -10,6 +10,16 @@ java {
     targetCompatibility = JavaVersion.VERSION_17
 }
 
+repositories {
+    mavenCentral()
+}
+
+configurations {
+    umlDoclet
+}
+
+apply plugin: 'java'
+
 def ROBOT_MAIN_CLASS = "frc.robot.Main"
 
 // Set this to true to enable desktop support.
@@ -35,6 +45,8 @@ dependencies {
 
     def akitJson = new groovy.json.JsonSlurper().parseText(new File(projectDir.getAbsolutePath() + "/vendordeps/AdvantageKit.json").text)
     annotationProcessor "org.littletonrobotics.akit:akit-autolog:$akitJson.version"
+
+    umlDoclet 'nl.talsmasoftware:umldoclet:2.2.1'
 }
 
 test {
@@ -153,3 +165,18 @@ spotless {
         endWithNewline()
     }
 }
+
+javadoc {
+    // exclude 'ascope_assets/**'
+    options {
+        links 'https://docs.oracle.com/en/java/javase/11/docs/api/', 'https://github.wpilib.org/allwpilib/docs/release/java', 'https://javadocs.photonvision.org/release', 'https://pathplanner.dev/api/java'
+        //, 'https://api.ctr-electronics.com/pheonix6/stable/java'
+        linkSource true
+        docletpath = configurations.umlDoclet.files.asType(List)
+        doclet = 'nl.talsmasoftware.umldoclet.UMLDoclet'
+        addBooleanOption 'createPumlFiles', true
+
+        // prevent failure on javadoc warnings
+        addStringOption('Xdoclint:none', '-quiet')
+    }
+}
